File: couchbase-entity/src/main/java/com/kaufland/generation/EntityGeneration.java
package com.kaufland.generation;

import com.kaufland.model.entity.EntityHolder;
import com.kaufland.model.field.CblBaseFieldHolder;
import com.kaufland.model.field.CblConstantHolder;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import java.util.Arrays;
import java.util.List;

import javax.lang.model.element.Modifier;

import kaufland.com.coachbasebinderapi.PersistenceConfig;
import kaufland.com.coachbasebinderapi.PersistenceException;

public class EntityGeneration {

    private static final String GET_DOCUMENT_METHOD = "getInstance().getConnector().getDocument";

    private static final String DELETE_DOCUMENT_METHOD = "getInstance().getConnector().deleteDocument";

    private static final String UPSERT_DOCUMENT_METHOD = "getInstance().getConnector().upsertDocument";


    public JavaFile generateModel(EntityHolder holder) {

        TypeSpec.Builder typeBuilder = TypeSpec.classBuilder(holder.getEntitySimpleName()).
                addModifiers(Modifier.PUBLIC).
                addField(CblDefaultGeneration.field()).
                addField(TypeUtil.createMapStringObject(), "mDoc", Modifier.PRIVATE).
                addField(TypeUtil.createMapStringObject(), "mDocChanges", Modifier.PRIVATE).
                addMethods(create(holder)).
                addMethod(contructor(holder)).
                addMethod(setAll(holder)).
                addMethods(new TypeConversionMethodsGeneration().generate()).
                addMethod(getId()).
                addMethod(toMap(holder)).
                addField(idConstant()).
                superclass(TypeName.get(holder.getSourceElement().asType()));


        for (CblBaseFieldHolder fieldHolder : holder.getAllFields()) {

            typeBuilder.addFields(fieldHolder.createFieldConstant());
            typeBuilder.addMethod(fieldHolder.getter(holder.getDbName(), true));

            MethodSpec setter = fieldHolder.setter(holder.getDbName(), holder.getEntityTypeName(), true);
            if (setter != null) {
                typeBuilder.addMethod(setter);
            }
        }

        typeBuilder.addStaticBlock(CblDefaultGeneration.staticInitialiser(holder));
        typeBuilder.addMethod(new RebindMethodGeneration().generate(true));
        typeBuilder.addMethod(delete(holder));
        typeBuilder.addMethod(save(holder));

        return JavaFile.builder(holder.getPackage(), typeBuilder.build()).
                build();

    }


    private FieldSpec idConstant() {
        return FieldSpec.builder(String.class, "_ID", Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).
                initializer("$S", "_id").
                build();
    }

    private MethodSpec getId() {

        return MethodSpec.methodBuilder("getId").
                addModifiers(Modifier.PUBLIC).
                returns(String.class).
                addStatement("return ($T) mDoc.get($N)", String.class, "_ID").
                build();
    }



    private MethodSpec setAll(EntityHolder holder) {

        MethodSpec.Builder setAllBuilder = MethodSpec.methodBuilder("setAll").addModifiers(Modifier.PUBLIC).
                addParameter(TypeUtil.createMapStringObject(), "map").
                addStatement("mDocChanges.putAll(map)", TypeUtil.createMapStringObject(), PersistenceConfig.class, holder.getDbName());

        return setAllBuilder.build();
    }

    private MethodSpec toMap(EntityHolder holder) {

        MethodSpec.Builder toMapBuilder = MethodSpec.methodBuilder("toMap").addModifiers(Modifier.PUBLIC).
                returns(TypeUtil.createMapStringObject()).
                addStatement("$T doc = $T." + GET_DOCUMENT_METHOD + "(getId(), $S)", TypeUtil.createMapStringObject(), PersistenceConfig.class, holder.getDbName());

        for (CblConstantHolder constantField : holder.getFieldConstants()) {
            toMapBuilder.addStatement("mDocChanges.put($S, $S)", constantField.getDbField(), constantField.getConstantValue());
        }

        toMapBuilder.addStatement("$1T temp = new $1T()", TypeUtil.createHashMapStringObject());
        toMapBuilder.addCode(CodeBlock.builder().
                addStatement("temp.putAll(mDocDefaults)").
                beginControlFlow("if(doc != null)").
                addStatement("temp.putAll(doc)").
                endControlFlow().
                beginControlFlow("if(mDocChanges != null)").
                addStatement("temp.putAll(mDocChanges)").
                endControlFlow().
                addStatement("return temp").
                build());

        return toMapBuilder.build();
    }

    private MethodSpec delete(EntityHolder holder) {
        return MethodSpec.methodBuilder("delete").addModifiers(Modifier.PUBLIC).
                addException(PersistenceException.class).
                addStatement("$N." + DELETE_DOCUMENT_METHOD + "(getId(), $S)", PersistenceConfig.class.getCanonicalName(), holder.getDbName()).
                build();
    }

    private MethodSpec save(EntityHolder holder) {
        MethodSpec.Builder saveBuilder = MethodSpec.methodBuilder("save").addModifiers(Modifier.PUBLIC).
                addException(PersistenceException.class).
                addStatement("$T doc = toMap()", TypeUtil.createMapStringObject());

        saveBuilder.addStatement("$N." + UPSERT_DOCUMENT_METHOD + "(doc, getId(), $S)", PersistenceConfig.class.getCanonicalName(), holder.getDbName());
        saveBuilder.addStatement("rebind(doc)");

        return saveBuilder.build();
    }

    private MethodSpec contructor(EntityHolder holder) {
        return MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC).
                addParameter(TypeUtil.createMapStringObject(), "doc").
                addStatement("rebind(doc)").
                build();
    }

    private List<MethodSpec> create(EntityHolder holder) {

        return Arrays.asList(
                MethodSpec.methodBuilder("create").
                        addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                        addParameter(String.class, "id").
                        addStatement("return new $N ($N." + GET_DOCUMENT_METHOD + "(id, $S))",
                                holder.getEntitySimpleName(), PersistenceConfig.class.getCanonicalName(), holder.getDbName()).
                        returns(holder.getEntityTypeName()).
                        build(),
                MethodSpec.methodBuilder("create").
                        addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                        addStatement("return new $N ($N." + GET_DOCUMENT_METHOD + "(null, $S))",
                                holder.getEntitySimpleName(), PersistenceConfig.class.getCanonicalName(), holder.getDbName()).
                        returns(holder.getEntityTypeName()).
                        build(),
                MethodSpec.methodBuilder("create").
                        addModifiers(Modifier.PUBLIC, Modifier.STATIC).
                        addParameter(TypeUtil.createMapStringObject(), "map").
                        addStatement("return new $N (map)",
                                holder.getEntitySimpleName()).
                        returns(holder.getEntityTypeName()).
                        build()
        );
    }

}

File: couchbase-entity/src/main/java/com/kaufland/generation/EntityGeneration.kt
package com.kaufland.generation

import com.kaufland.model.entity.EntityHolder
import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.jvm.throws
import kaufland.com.coachbasebinderapi.PersistenceConfig
import kaufland.com.coachbasebinderapi.PersistenceException
import java.util.*

class EntityGeneration {

    private val id: FunSpec
        get() = FunSpec.builder("getId").addModifiers(KModifier.PUBLIC).returns(String::class).addStatement("return mDoc.get(%N) as %T", "_ID", TypeUtil.string()).build()


    fun generateModel(holder: EntityHolder, useSuspend : Boolean): FileSpec {

        var companionSpec = TypeSpec.companionObjectBuilder()
        companionSpec.addProperty(idConstant())
        companionSpec.addFunctions(create(holder, useSuspend))
        companionSpec.addFunction(findById(holder, useSuspend))

        for (query in holder.queries) {
            query.queryFun(holder.dbName, holder, useSuspend)?.let {
                companionSpec.addFunction(it)
            }
        }

        for (generateAccessor in holder.generateAccessors) {
            generateAccessor.accessorFunSpec()?.let {
                companionSpec.addFunction(it)
            }
            generateAccessor.accessorPropertySpec()?.let {
                companionSpec.addProperty(it)
            }
        }

        val builderBuilder = BuilderClassGeneration.generateBaseBuilder(holder)

        val typeBuilder = TypeSpec.classBuilder(holder.entitySimpleName).addModifiers(KModifier.PUBLIC).addSuperinterface(TypeUtil.mapSupport())
                .addFunction(CblDefaultGeneration.addDefaults(holder))
                .addFunction(CblConstantGeneration.addConstants(holder))
                .addProperty(PropertySpec.builder("mDoc", TypeUtil.mutableMapStringObject(), KModifier.PRIVATE).mutable().initializer("%T()", TypeUtil.hashMapStringObject()).build())
                .addProperty(PropertySpec.builder("mDocChanges", TypeUtil.mutableMapStringObject(), KModifier.PRIVATE).mutable().initializer("%T()", TypeUtil.hashMapStringObject()).build())
                .addFunction(contructor(holder)).addFunction(setAll(holder)).addFunctions(TypeConversionMethodsGeneration(useSuspend).generate()).addFunction(id).superclass(holder.sourceElement!!.asType().asTypeName())
                .addFunction(toMap(holder, useSuspend))
                .addFunction(BuilderClassGeneration.generateBuilderFun())

        for (fieldHolder in holder.allFields) {

            fieldHolder.builderSetter(holder.dbName, holder.`package`, holder.entitySimpleName, true)?.let {
                builderBuilder.addFunction(it)
            }

            companionSpec.addProperties(fieldHolder.createFieldConstant())
            typeBuilder.addProperty(fieldHolder.property(holder.dbName, holder.abstractParts, true))
        }

        typeBuilder.addType(companionSpec.build())
        typeBuilder.addFunction(RebindMethodGeneration().generate(true))
        typeBuilder.addFunction(delete(holder, useSuspend))
        typeBuilder.addFunction(save(holder, useSuspend))

        typeBuilder.addType(builderBuilder.build())

        return FileSpec.get(holder.`package`, typeBuilder.build())

    }

    private fun findById(holder: EntityHolder, useSuspend: Boolean): FunSpec {
       return FunSpec.builder("findById").addModifiers(evaluateModifiers(useSuspend)).addParameter("id", String::class).addAnnotation(JvmStatic::class)
               .addStatement("val result = %T.${getDocumentMethod(useSuspend)}(id, %S)", PersistenceConfig::class, holder.dbName)
               .addStatement("return if(result != null) %N(result) else null", holder.entitySimpleName)
               .returns(holder.entityTypeName.copy(true)).build()
    }


    private fun idConstant(): PropertySpec {
        return PropertySpec.builder("_ID", String::class, KModifier.PUBLIC, KModifier.FINAL).initializer("%S", "_id").addAnnotation(JvmField::class).build()
    }


    private fun setAll(holder: EntityHolder): FunSpec {

        val setAllBuilder = FunSpec.builder("setAll").addModifiers(KModifier.PUBLIC).addParameter("map", TypeUtil.mapStringObject()).addStatement("mDocChanges.putAll(map)", TypeUtil.mapStringObject(), PersistenceConfig::class, holder.dbName)

        return setAllBuilder.build()
    }

    private fun toMap(holder: EntityHolder, useSuspend: Boolean): FunSpec {

        var refreshDoc = "%T.${getDocumentMethod(useSuspend)}(getId(), %S)"

        if(useSuspend){
            refreshDoc = "kotlinx.coroutines.runBlocking{$refreshDoc}"
        }

        val toMapBuilder = FunSpec.builder("toMap").addModifiers(KModifier.OVERRIDE).returns(TypeUtil.mutableMapStringObject()).addStatement("val doc = $refreshDoc", PersistenceConfig::class, holder.dbName)

        for (constantField in holder.fieldConstants.values) {
            toMapBuilder.addStatement("mDocChanges.put(%S, %S)", constantField.dbField, constantField.constantValue)
        }

        toMapBuilder.addStatement("var temp = %T()", TypeUtil.hashMapStringObject())
        toMapBuilder.addCode(CodeBlock.builder().beginControlFlow("if(doc != null)").addStatement("temp.putAll(doc)").endControlFlow().beginControlFlow("if(mDocChanges != null)").addStatement("temp.putAll(mDocChanges)").endControlFlow().addStatement("return temp").build())

        return toMapBuilder.build()
    }

    private fun delete(holder: EntityHolder, useSuspend: Boolean): FunSpec {
        return FunSpec.builder("delete").addModifiers(evaluateModifiers(useSuspend)).throws(PersistenceException::class).addStatement("%T.${deleteDocumentMethod(useSuspend)}(getId(), %S)", PersistenceConfig::class, holder.dbName).build()
    }

    private fun save(holder: EntityHolder, useSuspend: Boolean): FunSpec {
        val saveBuilder = FunSpec.builder("save").addModifiers(evaluateModifiers(useSuspend)).throws(PersistenceException::class).addStatement("val doc = toMap()")

        saveBuilder.addStatement("%T.${upsertDocumentMethod(useSuspend)}(doc, getId(), %S)", PersistenceConfig::class, holder.dbName)
        saveBuilder.addStatement("rebind(doc)")

        return saveBuilder.build()
    }

    private fun contructor(holder: EntityHolder): FunSpec {
        return FunSpec.constructorBuilder().addModifiers(KModifier.PUBLIC).addParameter("doc", TypeUtil.mapStringObject()).addStatement("rebind(doc)").build()
    }

    private fun evaluateModifiers(useSuspend: Boolean): List<KModifier> {
       return if(useSuspend) listOf(KModifier.PUBLIC, KModifier.SUSPEND) else listOf(KModifier.PUBLIC)
    }

    private fun create(holder: EntityHolder, useSuspend: Boolean): List<FunSpec> {

        return Arrays.asList(
                FunSpec.builder("create").addModifiers(evaluateModifiers(useSuspend)).addParameter("id", String::class).addAnnotation(JvmStatic::class).addStatement("return %N(%T.${getDocumentMethod(useSuspend)}(id, %S))",
                        holder.entitySimpleName, PersistenceConfig::class, holder.dbName).returns(holder.entityTypeName).build(),
                FunSpec.builder("create").addModifiers(evaluateModifiers(useSuspend)).addAnnotation(JvmStatic::class).addStatement("return %N(%T.${getDocumentMethod(useSuspend)}(null, %S))",
                        holder.entitySimpleName, PersistenceConfig::class, holder.dbName).returns(holder.entityTypeName).build(),
                FunSpec.builder("create").addModifiers(KModifier.PUBLIC).addParameter("map", TypeUtil.mutableMapStringObject()).addAnnotation(JvmStatic::class).addStatement("return %N(map)",
                        holder.entitySimpleName).returns(holder.entityTypeName).build()
        )
    }

    companion object {

        private fun getDocumentMethod(useSuspend: Boolean) : String{
            return "${if(useSuspend) "suspendingConnector" else "connector"}.getDocument"
        }

        private fun deleteDocumentMethod(useSuspend: Boolean) : String{
            return "${if (useSuspend) "suspendingConnector" else "connector"}.deleteDocument"
        }

        private fun upsertDocumentMethod(useSuspend: Boolean) : String{
            return "${if (useSuspend) "suspendingConnector" else "connector"}.upsertDocument"
        }
    }

}
